resctrl/utils.go

```diff
@@ -113,60 +116,80 @@ func prepareMonitoringGroup(containerName string, getContainerPids func() ([]str
-	if monGroupPath == "" {
-		// Remove leading prefix.
-		// e.g. /my/container -> my/container
-		if len(containerName) >= minContainerNameLen && containerName[0] == containerPrefix {
-			containerName = containerName[1:]
+	createNew := false
+
+	// Check if there is any monitoring group.
+	existingPath, err := findGroup(filepath.Join(controlGroupPath, monGroupsDirName), processThreads, false, true)
+	if err != nil {
+		if err != errNotEnoughPIDs && err != errTooManyPIDs {
+			return "", fmt.Errorf("couldn't find monitoring group matching %q container: %v", containerName, err)
 		}
 
-		// Add own prefix and use `-` instead `/`.
-		// e.g. my/container -> cadvisor-my-container
-		properContainerName := fmt.Sprintf("%s-%s", monGroupPrefix, strings.Replace(containerName, "/", "-", -1))
-		monGroupPath = filepath.Join(controlGroupPath, monitoringGroupDir, properContainerName)
+		rmErr := os.Remove(monGroupPath)
+		if rmErr != nil && !os.IsNotExist(rmErr) {
+			return "", fmt.Errorf("couldn't clean up monitoring group matching %q container: %v", containerName, rmErr)
+		}
+		if existingPath != monGroupPath {
+			rmErr = os.Remove(existingPath)
+			if rmErr != nil && !os.IsNotExist(rmErr) {
+				return "", fmt.Errorf("couldn't clean up monitoring group matching %q container: %v", containerName, rmErr)
+			}
+		}
+		createNew = true
+	}
 
+	// Prepare new one if not exists.
+	if createNew || existingPath == "" {
 		err = os.MkdirAll(monGroupPath, os.ModePerm)
 		if err != nil {
 			return "", fmt.Errorf("couldn't create monitoring group directory for %q container: %w", containerName, err)
 		}
-
-		if !inHostNamespace {
-			processPath = "/rootfs/proc"
-		}
-
-		for _, pid := range pids {
-			processThreads, err := getAllProcessThreads(filepath.Join(processPath, pid, processTask))
+		for _, thread := range processThreads {
+			treadInt, err := strconv.Atoi(thread)
 			if err != nil {
-				return "", err
+				return "", fmt.Errorf("couldn't parse %q: %w", thread, err)
 			}
-			for _, thread := range processThreads {
-				err = intelrdt.WriteIntelRdtTasks(monGroupPath, thread)
-				if err != nil {
-					secondError := os.Remove(monGroupPath)
-					if secondError != nil {
-						return "", fmt.Errorf(
-							"coudn't assign pids to %q container monitoring group: %w \n couldn't clear %q monitoring group: %v",
-							containerName, err, containerName, secondError)
-					}
-					return "", fmt.Errorf("coudn't assign pids to %q container monitoring group: %w", containerName, err)
+			err = intelrdt.WriteIntelRdtTasks(monGroupPath, treadInt)
+			if err != nil {
+				secondError := os.Remove(monGroupPath)
+				if secondError != nil {
+					return "", fmt.Errorf(
+						"coudn't assign pids to %q container monitoring group: %w \n couldn't clear %q monitoring group: %v",
+						containerName, err, containerName, secondError)
 				}
+				return "", fmt.Errorf("coudn't assign pids to %q container monitoring group: %w", containerName, err)
```

Comment:
Correct me if I'm wrong but:
* The Old Way:
  1. Iterate over all `pid`s. 
  2. Fetch all the `tid`s for a `pid`
  3. Iterate over all `tid`s 
  4. Write the `tid` to monitoring group.
  5. Exit `tid`s loop.
  6. Exit `pid`s loop.
* The New Way:
  1. Iterate over all `pid`s.
  2. Save all the `tid`s for a `pid` to a variable.
  3. Exit `pid`s loop.
  4. Iterate over all `tid`s.
  7. Write the `tid` to monitoring group.
  8. Exit `tid`s loop.

Functionally these two approaches are identical. In the PR description you wrote:

> Collect threads instead of pids. During collector startup we collect threads, so for the following checks we should also collect threads, not pids. Pid approach woks on relatively simple processes, but for any complex one pid list will never be the same as the original thread list, thus resulting in err on every collector run.

, but threads (`tid`s) have always been collected and written to monitoring group.

I might be missing something but I can't understand where the bug that you are trying to fix is. I will appreciate more detailed explanation that will help me to understand your reasoning. A test case failing with The Old Way and passing with The New Way would be perfect!